csharp/src/Ice/SignaledSocketStream.cs

```diff
@@ -13,9 +13,30 @@ namespace ZeroC.Ice
     /// for receiving data. The socket can easily signal the stream when new data is available.</summary>
     internal abstract class SignaledSocketStream<T> : SocketStream, IValueTaskSource<T>
     {
-        internal bool IsSignaled => Thread.VolatileRead(ref _signaled) == 1;
-        private volatile Exception? _exception;
-        private int _signaled;
+        internal bool IsSignaled
+        {
+            get
+            {
+                bool lockTaken = false;
+                try
+                {
+                    _lock.Enter(ref lockTaken);
+                    return _source.GetStatus(_source.Version) != ValueTaskSourceStatus.Pending;
```

Comment:
The parameter to GetStatus is an opaque token given to the ValueTask ctor. Why is it _source.Version?

Reply:
The way `SignalSocketStream` is used, there's always a single pending `WaitSignalAsync` call. So the version passed as the opaque token of ValueTask constructor should always be the same as the version of the `ManualResetValueTaskSourceCore` version.